### ISSUE TITLE
handling SIGTERMs when dealing with long requests 

### ISSUE THREAD

**Issue by everplays** [puma/puma#932]
Hi there,

##### TL;DR:

How can we receive SIGTERMs in the app itself and still let puma gracefully exit after request was processed?

##### Long version:

We are using Amazon SQS as backend of our job queue. Essentially, it runs the app just as it would run a web app instance but instead of directing normal traffic, it makes requests that are background jobs. Everything is working fine until we deploy or elastic beanstalk tries to scale. In either of the scenarios, they issue a SIGTERM to let the app finish what it is doing before SIGKILL but it's a long running request, it's meant to run for a few minutes! Therefore, puma doesn't get a chance to do a graceful stop as the execution is still in the app. What I am looking for is a way to communicate to Puma from within the app. Either of the following can work for us:

‌1. App asks puma if it has received a SIGTERM and if that was the case, stops what it's doing:

```ruby
def I_am_an_action_that_takes_5_minutes_to_run
  (1..300).each do
    sleep 1
    if env['puma.control'].status == :stop
      put_the_job_back_to_queue!
      break
    end
  end
end
```

‌2. App traps the signals (which prevents Puma from receiving them), handles the situation, and tells puma to stop:

```ruby
def I_am_an_action_that_takes_5_minutes_to_run
  (1..300).each do
    sleep 1
    SignalHandler.check!
  end
rescue SignalException => e
  put_the_job_back_to_queue!
end

# and SignalHandler.check! does something like the following in case of SIGTERM
if received_sigterm?
  # this is essential, otherwise, puma keeps running after request.
  env['puma.control'].set_status(:stop)
  raise SignalException, 15
end
```

Please let me know what you think and if you have a better suggestion. I would like to make a PR to address this.

**Comment by evanphx** [comment]
I've added a new feature called `force_shutdown_after` that allows you to tell Puma how long to wait for threads to finish their work before being forced to stop. Would that solve your problem?

**Comment by everplays** [comment]
Thanks for looking into this @evanphx. `force_shutdown_after` could make the situation better but will not solve it entirely.

**What it can improve:**
We split the background jobs and try to keep them under the time that beanstalk waits for the server. For the most of the jobs, it is possible and `force_shutdown_after` solves the problem.

**The problem which remains:**
Imagine that we use `forever` for `force_shutdown_after` (which is the default value based on the code) and we do something that potentially could take 15 minutes to finish. After the initial SIGTERM, beanstalk will not wait 15 minutes but will issue a SIGKILL and terminates the process all together. In this scenario, `force_shutdown_after` will not solve the issue.

**Comment by evanphx** [comment]
I think you're going to better off using beanstalk or sidekiq for this. Trying to this kind of very long request in a request oriented server like Puma is always going to be painful.

**Comment by everplays** [comment]
I absolutely agree with you. However, if we want to keep in the standard realm of elastic beanstalk, we can not run our own daemons. Therefore, using something like Sidekiq or DelayedJob is not possible.

**Comment by evanphx** [comment]
You can run Sidekiq in the same process of Puma, if that helps.

**Comment by everplays** [comment]
@evanphx sorry for the delay. I was checking if it would solve the problem but unfortunately it doesn't. Still there is no way to determine if puma has received a `SIGTERM` in the middle of executing a job.

If you are willing to expose some kind of API to allow application get this information from puma, I would like to work on it.

**Comment by evanphx** [comment]
@everplays I could set some flag that indicates puma is trying to shutdown, your app could check the flag and take actions. Would that work?

**Comment by everplays** [comment]
@evanphx yes, that will work. We can stop whatever we are doing when the flag changes and let puma handle the signal by itself.
(It will be comparable with the first code snippet that I put in the description)

**Comment by evanphx** [comment]
@everplays Ok, a new api `Puma::Server.current.shutting_down?` can now be used to detect if the server is trying to shutdown.

**Comment by everplays** [comment]
@evanphx perfect. Thanks for taking care of it.
